nova/virt/vmwareapi/vmops.py

```diff
@@ -2232,3 +2247,88 @@ def get_mks_console(self, instance):
                     'thumbprint': thumbprint}
         internal_access_path = jsonutils.dumps(mks_auth)
         return ctype.ConsoleMKS(ticket.host, ticket.port, internal_access_path)
+
+    @synchronized('update_cache')
+    def update_cached_instances(self):
+        vim = self._session.vim
+        options = None
+        if self._property_collector is None:
+            self._property_collector = vim.service_content.propertyCollector
+            vim.CreateFilter(self._property_collector, spec=self._get_vm_monitor_spec(vim), partialUpdates=False)
+        update_set = vim.WaitForUpdatesEx(self._property_collector, version="",
+                                          options=options)
+        while update_set:
+
+            self._property_collector_version = update_set.version
+            if update_set.filterSet and update_set.filterSet[0].objectSet:
+                for update in update_set.filterSet[0].objectSet:
+                    if update.obj['_type'] == "VirtualMachine":
+                        if update.kind == "leave":
+                            LOG.info("Removing instance from cache...")
+                            vm_refs_cache = copy.copy(vm_util._VM_REFS_CACHE)
+                            inv_cache_map = {v.value: k for k, v in vm_refs_cache.items()}
+                            cache_id_to_delete = inv_cache_map.get(update.obj.value)
```

Comment:
Here you can replace inv_cache_map with
`cache_id_to_delete = vm_util._VM_VALUE_CACHE.get(update.obj.value, {}).get('config.instanceUuid')`
(When the `_VM_VALUE_CACHE` is keyed by `vm_ref`)